pricegraph/src/graph/shortest_paths.rs

```diff
@@ -81,69 +104,70 @@ where
+    fn path_to(&self, dest: G::NodeId) -> Option<Path<G::NodeId>> {
         self.predecessor_store
             .path_to(self.source, dest, self.graph)
     }
 
     /// Lists all nodes that can be reached from the source (including the source itself).
-    pub fn connected_nodes(&self) -> Vec<G::NodeId> {
+    fn connected_nodes(&self) -> Vec<G::NodeId> {
         let mut node_indices = self.predecessor_store.connected_nodes(self.graph);
         debug_assert!(!node_indices.contains(&self.source));
         node_indices.push(self.source);
         node_indices
     }
 
-    /// Initializes a shortest path graph that will be later built with the
-    /// Bellman-Ford algorithm.
-    fn empty(g: G, source: G::NodeId, hops: Option<usize>) -> Self {
-        let predecessors = vec![None; g.node_bound()];
-        let mut distances = vec![<_>::infinite(); g.node_bound()];
-        distances[g.to_index(source)] = <_>::zero();
-
-        let predecessor_store: Box<dyn PredecessorStoring<G>> = match hops {
-            None => Box::new(Unbounded::new(predecessors, distances)),
-            Some(h) => Box::new(Bounded::new(predecessors, distances, h))
-        };
-
-        ShortestPathGraph {
-            graph: g,
-            predecessor_store,
-            source,
-        }
-    }
-
     fn prepare_next_relaxation_step(&mut self) {
         self.predecessor_store.prepare_next_relaxation_step();
     }
 
-    /// Checks for negative weight cycle and, if any is found, creates loop in
-    /// the predecessor store
-    fn mark_cycle(&mut self) -> Option<G::NodeId> {
-        self.predecessor_store.mark_cycle(self.graph)
-    }
-
     /// Returns a negative cycle, if it exists.
     fn find_cycle(&mut self) -> Option<NegativeCycle<G::NodeId>> {
-        let search_start = match self.mark_cycle() {
+        let search_start = match self.predecessor_store.mark_cycle(self.graph) {
             Some(node) => node,
             None => return None,
         };
         self.predecessor_store.find_cycle(search_start, self.graph)
     }
+}
 
-    /// Creates a representation of all shortest paths from the given source
-    /// to any other node in the graph.
-    ///
-    /// Shortest paths are well defined if and only if the graph does not
-    /// contain any negative weight cycle reachabe from the source. If a
-    /// negative weight cycle is detected, it is returned as an error.
-    pub fn new(
-        g: G,
-        source: G::NodeId,
-        hops: Option<usize>,
-    ) -> Result<Self, NegativeCycle<G::NodeId>> {
-        bellman_ford(g, source, hops)
+/// Creates a representation of all shortest paths from the given source
+/// to any other node in the graph.
+///
+/// Shortest paths are well defined if and only if the graph does not
+/// contain any negative weight cycle reachabe from the source. If a
+/// negative weight cycle is detected, it is returned as an error.
+pub fn new_shortest_path_graph<'a, G>(
```

Comment:
ubernit: Maybe this should be called
```suggestion
pub fn shortest_path<'a, G>(
```